### PR TITLE
Link analysis

### PR DESCRIPTION
**Pull request by khattakdev**
- [x] Works for Run Analysis button 
run Analysis

#89

### CODE REVIEW

File: src/components/MapView/Analyser/index.tsx
Comment (ericboucher) [inline src/components/MapView/Analyser/index.tsx:403]: is that change necessary?

File: .gitignore
Comment (ericboucher) [inline .gitignore:35]: is this needed?

**Review by ericboucher — COMMENTED** [review]
This is a great start. I would recommend moving the URL handling into a separate file to keep things clean and think about how you can DRY your code. a bit more.

File: src/components/MapView/Analyser/index.tsx
Comment (ericboucher) [inline src/components/MapView/Analyser/index.tsx:196]: Let's create a smarter, DRYer util function for that and make sure we handle edge cases. What happens when some values are undefined? eg. for thresholds?

**Comment by ericboucher** [comment]
Some tests are failing, I think you just need to mock a few more functions around `history`

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:217]:
```suggestion
    function isBaselineLayerId(layer: string): layer is LayerKey {
```

Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:217]: If you are stuck on how to complete these functions, I recommend checking the key exists here: https://github.com/WFP-VAM/prism-frontend/blob/master/src/config/utils.ts#L116, then checking the `layer.type` to verify its the correct type (Hazard - WMS or Baseline - NSO)

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:103]: I don't think this is a good idea. The goal of this useEffect is to load default dates when a new hazard layer is selected. We will now disable this feature if there's something in the url.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:229]: Only mean is supported in the url?

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:242]: We shouldn't aim for this code to just be run on startup. If the url changes because it is manually edited, or the user goes back a page, the page should change accordingly.

**Review by Ry-DS — CHANGES_REQUESTED** [review]

**Comment by Ry-DS** [comment]
Good start! However, I think we might have to change our approach on how we load and update the url.

The goal of this PR is to help users easily share *completed* analysis results (all fields are filled in and run analysis was dispatched). There is a lot of extra work we are doing here to try handle all these smaller edge cases where certain fields are not filled in, and the code is also not very scalable if we decide to add new fields in the future.

I think its safe if we just ignore filling in the UI's fields, and just dispatch an analysis if all criteria are met. If a dispatch isn't possible, we could simply notify the user that their url is malformed. A url would only ever be malformed if the user edited it, or didn't copy it properly. 

The above approach would definitely simplify things, and you won't need to touch some of the more delicate pieces like https://github.com/WFP-VAM/prism-frontend/pull/120#discussion_r551077956. What do you think of this approach @ericboucher? This is the current approach #103 was taking.

Also, I don't think we should be doing any verification when we load the url like what we are doing now, but instead rely on the dispatch function to verify for us. Then, in the future we could use this function in other places and use the same verification code. This was the approach Laveesh and I were taking [here.](https://github.com/WFP-VAM/prism-frontend/pull/103/files#diff-edfb4dd8e4a2f9b60a39a96700305dadc49e358206c8f5b88f82343e67529c73R45)

Try give the above a shot, and possibly use #103 as a jumping point. Let me know if you have any questions :+1:

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:403]: It wasn't working initially, so I had to use an anonymous function which fixed the issue but reverted it back just now and it worked fine.

File: .gitignore
Comment (khattakdev) [reply .gitignore:35]: There is a VSCODE extension that you can use for adding **TODOs** so this was added intentionally to ignore `Todo`

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:217]: > If you are stuck on how to complete these functions, I recommend checking the key exists here: https://github.com/WFP-VAM/prism-frontend/blob/master/src/config/utils.ts#L116, then checking the `layer.type` to verify its the correct type (Hazard - WMS or Baseline - NSO)

The function is already complete, It's doing what I wanted that is to check whether the passed argument is a correct Layer ( `"ndvi" | "ModisLST" | "ModisVHI" ... `)

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:103]: The goal of this if statement is to check if the `URL`  has a date in it, ignore the code within the `useEffect` which actually mess up with the URL dates and you don't get the expected result.

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:229]:
```
const [statistic, setStatistic] = useState(AggregationOperations.Mean);
```
This means `Mean` will always be selected by default, so we don't need to worry about it.
Statistics can have only two values either `Mean`  or `Median`, one is selected by default when declaring the value, we only need to care about the other one.

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:242]: This takes care of all the scenarios you mentioned.
If the user manually edited the URL, the page is going to rerender everything and that will execute `runAnalyser`. The same thing happens when users navigate backs.

**Comment by khattakdev** [comment]
> Good start! However, I think we might have to change our approach on how we load and update the url.
> 
> The goal of this PR is to help users easily share _completed_ analysis results (all fields are filled in and run analysis was dispatched). There is a lot of extra work we are doing here to try handle all these smaller edge cases where certain fields are not filled in, and the code is also not very scalable if we decide to add new fields in the future.
> 
> I think its safe if we just ignore filling in the UI's fields, and just dispatch an analysis if all criteria are met. If a dispatch isn't possible, we could simply notify the user that their url is malformed. A url would only ever be malformed if the user edited it, or didn't copy it properly - two cases we don't need to cover as developers.
> 
> The above approach would definitely simplify things, and you won't need to touch some of the more delicate pieces like [#120 (comment)](https://github.com/WFP-VAM/prism-frontend/pull/120#discussion_r551077956). What do you think of this approach @ericboucher? This is the current approach #103 was taking.
> 
> Also, I don't think we should be doing any verification when we load the url like what we are doing now, but instead rely on the dispatch function to verify for us. Then, in the future we could use this function in other places and use the same verification code. This was the approach Laveesh and I were taking [here.](https://github.com/WFP-VAM/prism-frontend/pull/103/files#diff-edfb4dd8e4a2f9b60a39a96700305dadc49e358206c8f5b88f82343e67529c73R45)
> 
> Try give the above a shot, and let me know if you have any questions 👍

Two reasons why I chose this approach.
* If a user opens such a link, he will have all the UI fields filled in, which can improve UX, Image if he wants to make a small change like changing the `Statistics` only, He will need to fill all the fields manually just for one change.

* Doing so, I don't need to make any changes in `runAnalyser` because it takes the values from input fields, all I need to do is to just update those fields and execute `runAnalyser`. 

```
 const params: AnalysisDispatchParams = {
      hazardLayer: selectedHazardLayer,
      baselineLayer: selectedBaselineLayer,
      date: selectedDate,
      statistic,
      extent: adminBoundariesExtent,
      threshold: {
        above: parseFloat(aboveThreshold) || undefined,
        below: parseFloat(belowThreshold) || undefined,
      },
    };
    
    await dispatch(requestAndStoreAnalysis(params));
```

This function's ( `requestAndStoreAnalysis` ) params can take specific values only, e.g. `hazardLayer` will only take valid layer names else TypeScript will return an error, so I'll still need to verify the user inputs to make sure it's not an unknown value.

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:196]: > Let's create a smarter, DRYer util function for that and make sure we handle edge cases. What happens when some values are undefined? eg. for thresholds?

I'll refactor the code once I get approval of my changes. 
Taking about the edge cases, I'm taking care of them already

```
    if (isHazardLayerId(hazardLayerParamId)) {
      setHazardLayerId(hazardLayerParamId);
    }
    if (isBaselineLayerId(baselineLayerParamId)) {
      setBaselineLayerId(baselineLayerParamId);
    }

    // Mean is selected by default
    if (statisticParam === 'median') {
      setStatistic(AggregationOperations.Median);
    }
```

Thresholds are optional, so I didn't add any checks for them. Also, they become undefined if they have an invalid number 
```
      threshold: {
        above: parseFloat(aboveThreshold) || undefined,
        below: parseFloat(belowThreshold) || undefined,
      },
```

File: .gitignore
Comment (Ry-DS) [reply .gitignore:35]: A comment then like  `# VSCode extensions` might be nice

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:103]: We still need to find a better way to do this. 

If I run an analysis, and copy paste the url to a new page, then try editing the hazard layer field, the dates won't update like they're supposed to. The approach I mentioned below (https://github.com/WFP-VAM/prism-frontend/pull/120#issuecomment-753710241) could work here. Otherwise, if there's a solution you can think of which still keeps this feature working, feel free to include it :+1:

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:229]: But if we rely on this default, things can easily break down the line. What if:

- We change our default?
- We add more options like `Mode`?
- The user triggers a re-run of this function while `Median` is selected but the url states `Mean`?

These cases will easily break this logic, therefore it would be great to not rely on defaults, especially in a useEffect.
If we also want to add Back support, where the user can navigate backwards, point 3 will come into play here.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:242]: I tried navigating back just now, but nothing happened. Which browser did you test on? (I used Edge)

We shouldn't ignore this eslint rule anyway, as its a paradigm of `useEffect` to exhaust all dependencies in use. Adding all dependencies will also add back support for all browsers.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:217]: Are you sure checking for a string type is good enough? We could pass absolutely anything like `not_a_layer` and cause a nasty error somewhere down the line. I tried just now, and it crashed the application without a graceful error/warning.

I recommend checking if the string is actually valid using the link I sent so we can verify the layer actually exists, and follow with a non-action or a graceful error.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [reply src/components/MapView/Analyser/index.tsx:103]: One I can think of off the top of my head is using a new boolean state to dictate if the state was set by the user or by the computer. If it was set by the computer, we can ignore this useEffect, otherwise we should still run this code. Hopefully this is helpful enough as a jumping point

**Comment by Ry-DS** [comment]
> Two reasons why I chose this approach.
> 
> * If a user opens such a link, he will have all the UI fields filled in, which can improve UX, Image if he wants to make a small change like changing the `Statistics` only, He will need to fill all the fields manually just for one change.
> * Doing so, I don't need to make any changes in `runAnalyser` because it takes the values from input fields, all I need to do is to just update those fields and execute `runAnalyser`.

This sounds fair, and filling all UI fields would be nice. Then, we should focus on the review comments I made, and making this more stable. 

For example, this might be a good place to use `useReducer`? Then we have the object keys available to us in order to construct a url, and if a new field is added it will be very easy to include it into the url without making a new condition for it (it would almost be automatic!) We could start this by removing all the `useState` uses related to the AnalysisUI with a reducer function, and `runAnalysis` can become an action. Loading a url could also be an action.

**Comment by khattakdev** [comment]
I've fixed pretty much all the issues, let me know if I missed any.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:106]: Remove this comment? Also, did you test if the `date` url query works after removing this? We'll need a solution that both allows this `useEffect` to work alongside the url query code. If you need help coming up with a way to do this let me know :+1:

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:114]: Where did this extra dependency come from?

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:200]: We need to find a better way than hard coding all the fields into the url. After talking with Eric, `useReducer` seems like a perfect fit here, as it will give you access to the object key names, so the url will automatically add new fields as they are introduced without any extra work. Let me know if you need help implementing this.

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:233]: I'd prefer if we'd use a try/catch so we can reuse the existing verification inside `runAnalyser()`

File: src/components/MapView/Analyser/index.tsx
Comment (Ry-DS) [inline src/components/MapView/Analyser/index.tsx:236]: You should be able to remove this, and use all dependencies ESLint asks you to include. Let me know if you face issues with this :+1: (we want this code to be written in a way that's independent of the current state of the application, so it remains stable. This eslint rule will help ensure this.)

File: src/components/MapView/Analyser/util.ts
Comment (Ry-DS) [inline src/components/MapView/Analyser/util.ts:1]: Move this into an action along with most the code in your `useEffect` once you start using `useReducer`

File: src/components/NavBar/index.tsx
Comment (Ry-DS) [inline src/components/NavBar/index.tsx:20]: What is this comment referring to?

File: src/components/NavBar/index.tsx
Comment (Ry-DS) [inline src/components/NavBar/index.tsx:73]: We really shouldn't need to do this. What if we navigate to `/` somewhere else in the codebase? We would need to include this function everywhere. 

A better way to do this would be to rely on the `useEffect` to clear the pending analysis if it detects the url has changed to `/`. This should happen automatically [here](https://github.com/WFP-VAM/prism-frontend/pull/120/files#diff-314a2832937cda41c6f51176b41683977e412b5dc6b10d8ffff85ffa2f293bd1R204)

File: src/config/types.ts
Comment (Ry-DS) [inline src/config/types.ts:215]: We could have a function for this, but I personally feel `stat in AggregationOptions` is readable enough, and I'm pretty sure TS is smart enough to apply its own type predicate if it sees this. This is up to discussion however, and you are free to keep this if you think its better :+1:

File: src/components/NavBar/index.test.tsx
Comment (Ry-DS) [inline src/components/NavBar/index.test.tsx:11]: This should be undone as per the comment I made on NavBar/index.tsx

**Review by Ry-DS — CHANGES_REQUESTED** [review]
I recommended some changes to help make this code more scalable, as right now it is quite restricted to our current use case.

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:233]: That will always return throw an error. The point of `if` statement is to avoid executing function at all. Before the values are updated, it tries to execute `runAnalyser` (and the layers are not defined yet).

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:114]: An extra dependency I added for dates but forgot to remove it. Nvm.

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:106]: Removed!

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:236]: In that case, I'll have to add `hazardLayerId` `baselineLayerId` `selectedDate` etc as well which will cause some issues. (If you do that, you won't be able to change the Analysis from UI, it will go into some infinity loop)

File: src/components/MapView/Analyser/index.tsx
Comment (khattakdev) [reply src/components/MapView/Analyser/index.tsx:200]: @Ry-DS Do you mean to have a reducer just for the URL which updates on every input field update?

**Pull request closed without merging**
